File: app/src/main/java/org/zeroxlab/momodict/db/realm/RealmStore.java
package org.zeroxlab.momodict.db.realm;

import android.content.Context;
import android.support.annotation.NonNull;
import android.text.TextUtils;

import org.zeroxlab.momodict.model.Book;
import org.zeroxlab.momodict.model.Card;
import org.zeroxlab.momodict.model.Entry;
import org.zeroxlab.momodict.model.Record;
import org.zeroxlab.momodict.model.Store;

import java.util.ArrayList;
import java.util.List;

import io.realm.Realm;
import io.realm.RealmResults;

public class RealmStore implements Store {

    private final Context mCtx;

    public RealmStore(Context ctx) {
        mCtx = ctx;
        Realm.init(mCtx);
    }

    @Override
    public boolean addBook(Book book) {
        final Realm realm = Realm.getDefaultInstance();
        realm.beginTransaction();
        final RealmBook managedDic =
                realm.createObject(RealmBook.class, book.bookName);
        managedDic.author = book.author;
        managedDic.wordCount = book.wordCount;
        managedDic.date = book.date;
        realm.commitTransaction();
        realm.close();
        return true;
    }

    @Override
    public Book getBook(String name) {
        throw new RuntimeException("Not implemented yet");
    }

    @Override
    public List<Book> getBooks() {
        final Realm realm = Realm.getDefaultInstance();
        final RealmResults<RealmBook> results =
                realm.where(RealmBook.class).findAll();
        final List<Book> dics = new ArrayList<>();
        for (RealmBook managedDic : results) {
            Book dic = new Book();
            dic.bookName = managedDic.bookName;
            dic.author = managedDic.author;
            dic.wordCount = managedDic.wordCount;
            dic.date = managedDic.date;
            dics.add(dic);
        }
        realm.close();
        return dics;
    }

    @Override
    public boolean removeBook(String name) {
        throw new RuntimeException("Not implemented yet");
    }

    @Override
    public boolean addEntries(List<Entry> entries) {
        final Realm realm = Realm.getDefaultInstance();
        realm.beginTransaction();
        for (Entry entry : entries) {
            RealmEntry managedEntry = realm.createObject(RealmEntry.class);
            managedEntry.wordStr = entry.wordStr;
            managedEntry.sourceBook = entry.source;
            managedEntry.data = entry.data;
        }
        realm.commitTransaction();
        realm.close();
        return true;
    }

    @Override
    public List<Entry> queryEntries(String keyWord) {
        final Realm realm = Realm.getDefaultInstance();
        final RealmResults<RealmEntry> managedEntries = TextUtils.isEmpty(keyWord)
                ? realm.where(RealmEntry.class).findAll()
                : realm.where(RealmEntry.class).contains("wordStr", keyWord).findAll();

        List<Entry> mapped = map(managedEntries);
        realm.close();
        return mapped;
    }

    @Override
    public List<Entry> getEntries(String keyWord) {
        final Realm realm = Realm.getDefaultInstance();
        if (TextUtils.isEmpty(keyWord)) {
            throw new RuntimeException("Keyword is empty");
        }
        final RealmResults<RealmEntry> managedEntries = realm
                .where(RealmEntry.class)
                .equalTo("wordStr", keyWord)
                .findAll();

        List<Entry> mapped = map(managedEntries);
        realm.close();
        return mapped;
    }

    @Override
    public boolean upsertRecord(@NonNull Record record) {
        final Realm realm = Realm.getDefaultInstance();
        RealmRecord previous = realm.where(RealmRecord.class)
                .equalTo("wordStr", record.wordStr)
                .findFirst();
        realm.beginTransaction();
        RealmRecord managedRecord = (previous == null)
                ? realm.createObject(RealmRecord.class, record.wordStr)
                : previous;
        managedRecord.count = record.count;
        managedRecord.time = record.time;
        realm.commitTransaction();
        realm.close();
        return true;
    }

    @Override
    public List<Record> getRecords() {
        final Realm realm = Realm.getDefaultInstance();
        final RealmResults<RealmRecord> managedRecords = realm
                .where(RealmRecord.class)
                .findAll();

        final List<Record> records = new ArrayList<>();
        for (int i = 0; i < managedRecords.size(); i++) {
            RealmRecord managedRecord = managedRecords.get(i);
            Record record = new Record();
            record.count = managedRecord.count;
            record.wordStr = managedRecord.wordStr;
            record.time = managedRecord.time;
            records.add(record);
        }
        realm.close();
        return records;
    }

    @Override
    public boolean upsertCard(@NonNull Card card) {
        final Realm realm = Realm.getDefaultInstance();
        RealmCard previous = realm.where(RealmCard.class)
                .equalTo("wordStr", card.wordStr)
                .findFirst();
        realm.beginTransaction();
        RealmCard managedCard = (previous == null)
                ? realm.createObject(RealmCard.class, card.wordStr)
                : previous;
        managedCard.time = card.time;
        managedCard.note = card.note;
        realm.commitTransaction();
        realm.close();
        return true;
    }

    @Override
    public boolean removeCards(@NonNull String keyWord) {
        final Realm realm = Realm.getDefaultInstance();
        realm.beginTransaction();
        RealmResults<RealmCard> rows = realm.where(RealmCard.class)
                .equalTo("wordStr", keyWord)
                .findAll();
        rows.deleteAllFromRealm();
        realm.commitTransaction();
        realm.close();
        return true;
    }

    @Override
    public List<Card> getCards() {
        final Realm realm = Realm.getDefaultInstance();
        final RealmResults<RealmCard> managedCards = realm
                .where(RealmCard.class)
                .findAll();

        final List<Card> cards = new ArrayList<>();
        for (int i = 0; i < managedCards.size(); i++) {
            RealmCard managedCard = managedCards.get(i);
            Card card = new Card();
            card.wordStr = managedCard.wordStr;
            card.note = managedCard.note;
            card.time = managedCard.time;
            cards.add(card);
        }
        realm.close();
        return cards;
    }

    @Override
    public boolean removeRecords(String keyWord) {
        final Realm realm = Realm.getDefaultInstance();
        realm.beginTransaction();
        RealmResults<RealmRecord> rows = realm.where(RealmRecord.class)
                .equalTo("wordStr", keyWord)
                .findAll();
        rows.deleteAllFromRealm();
        realm.commitTransaction();
        realm.close();
        return true;
    }

    @Override
    public List<Entry> queryEntries(String keyWord, String bookName) {
        throw new RuntimeException("Not implemented yet");
    }

    private List<Entry> map(RealmResults<RealmEntry> managedEntries) {
        final List<Entry> entries = new ArrayList<>();
        for (int i = 0; entries.size() < MAX_LENGTH && i < managedEntries.size(); i++) {
            RealmEntry managedEntry = managedEntries.get(i);
            Entry entry = new Entry();
            entry.source = managedEntry.sourceBook;
            entry.wordStr = managedEntry.wordStr;
            entry.data = managedEntry.data;
            entries.add(entry);
        }
        return entries;
    }
}

File: app/src/main/java/org/zeroxlab/momodict/db/realm/RealmStore.kt
package org.zeroxlab.momodict.db.realm

import android.content.Context
import android.os.StrictMode
import android.text.TextUtils
import io.realm.Realm
import io.realm.RealmResults
import org.zeroxlab.momodict.model.Book
import org.zeroxlab.momodict.model.Card
import org.zeroxlab.momodict.model.Entry
import org.zeroxlab.momodict.model.Record
import org.zeroxlab.momodict.model.Store

class RealmStore(private val mCtx: Context) : Store {

    init {
        // Realm need to load native .so files, turn off StrictMode for it
        val policy = StrictMode.getThreadPolicy()
        StrictMode.allowThreadDiskReads()
        Realm.init(mCtx)
        StrictMode.setThreadPolicy(policy)
    }

    override fun addBook(book: Book): Boolean {
        val realm = Realm.getDefaultInstance()
        realm.beginTransaction()
        val managedDic = realm.createObject(RealmBook::class.java, book.bookName)
        managedDic.author = book.author
        managedDic.wordCount = book.wordCount
        managedDic.date = book.date
        realm.commitTransaction()
        realm.close()
        return true
    }

    override fun getBook(name: String): Book {
        throw RuntimeException("Not implemented yet")
    }

    override fun getBooks(): List<Book> {
        val realm = Realm.getDefaultInstance()
        val dics = realm.where(RealmBook::class.java)
                .findAll()
                .map { managedDic ->
                    Book().also {
                        it.bookName = managedDic.bookName
                        it.author = managedDic.author
                        it.wordCount = managedDic.wordCount
                        it.date = managedDic.date
                    }
                }

        realm.close()
        return dics
    }

    override fun removeBook(bookName: String): Boolean {
        val openedRealm = Realm.getDefaultInstance().also { realm ->
            realm.beginTransaction()
            realm.where(RealmEntry::class.java)
                    .equalTo("sourceBook", bookName)
                    .findAll()
                    .deleteAllFromRealm()
            realm.where(RealmBook::class.java)
                    .equalTo("bookName", bookName)
                    .findAll()
                    .deleteAllFromRealm()
            realm.commitTransaction()
        }
        openedRealm.close()
        return true
    }

    override fun addEntries(entries: List<Entry>): Boolean {
        val realm = Realm.getDefaultInstance()

        realm.beginTransaction()
        entries.forEach { entry ->
            realm.createObject(RealmEntry::class.java).also { managedEntry ->
                managedEntry.wordStr = entry.wordStr
                managedEntry.sourceBook = entry.source
                managedEntry.data = entry.data
            }

        }
        realm.commitTransaction()
        realm.close()
        return true
    }

    override fun queryEntries(keyWord: String?): List<Entry> {
        val realm = Realm.getDefaultInstance()
        val managedEntries = if (TextUtils.isEmpty(keyWord))
            realm.where(RealmEntry::class.java).findAll()
        else
            realm.where(RealmEntry::class.java).contains("wordStr", keyWord).findAll()

        val mapped = map(managedEntries)
        realm.close()
        return mapped
    }

    override fun getEntries(keyWord: String?): List<Entry> {
        val realm = Realm.getDefaultInstance()
        if (TextUtils.isEmpty(keyWord)) {
            throw RuntimeException("Keyword is empty")
        }
        val managedEntries = realm
                .where(RealmEntry::class.java)
                .equalTo("wordStr", keyWord)
                .findAll()

        val mapped = map(managedEntries)
        realm.close()
        return mapped
    }

    override fun upsertRecord(record: Record): Boolean {
        val realm = Realm.getDefaultInstance()
        val previous = realm.where(RealmRecord::class.java)
                .equalTo("wordStr", record.wordStr)
                .findFirst()
        realm.beginTransaction()
        val managedRecord = previous ?: realm.createObject(RealmRecord::class.java, record.wordStr)
        managedRecord.count = record.count
        managedRecord.time = record.time
        realm.commitTransaction()
        realm.close()
        return true
    }

    override fun getRecords(): List<Record> {
        val realm = Realm.getDefaultInstance()
        val records = realm
                .where(RealmRecord::class.java)
                .findAll()
                .map { managedRecord ->
                    Record().also {
                        it.count = managedRecord.count
                        it.wordStr = managedRecord.wordStr
                        it.time = managedRecord.time
                    }
                }
        realm.close()
        return records
    }

    override fun upsertCard(card: Card): Boolean {
        val realm = Realm.getDefaultInstance()
        val previous = realm.where(RealmCard::class.java)
                .equalTo("wordStr", card.wordStr)
                .findFirst()
        realm.beginTransaction()
        val managedCard = previous ?: realm.createObject(RealmCard::class.java, card.wordStr)
        managedCard.time = card.time
        managedCard.note = card.note
        realm.commitTransaction()
        realm.close()
        return true
    }

    override fun removeCards(keyWord: String): Boolean {
        val realm = Realm.getDefaultInstance()
        realm.beginTransaction()
        val rows = realm.where(RealmCard::class.java)
                .equalTo("wordStr", keyWord)
                .findAll()
        rows.deleteAllFromRealm()
        realm.commitTransaction()
        realm.close()
        return true
    }

    override fun getCards(): List<Card> {
        val realm = Realm.getDefaultInstance()
        val managedCards = realm
                .where(RealmCard::class.java)
                .findAll()

        val cards = ArrayList<Card>()
        for (i in managedCards.indices) {
            val managedCard = managedCards[i]
            val card = Card()
            // FIXME: should remove !!
            card.wordStr = managedCard!!.wordStr
            card.note = managedCard.note
            card.time = managedCard.time
            cards.add(card)
        }
        realm.close()
        return cards
    }

    override fun removeRecords(keyWord: String): Boolean {
        val realm = Realm.getDefaultInstance()
        realm.beginTransaction()
        val rows = realm.where(RealmRecord::class.java)
                .equalTo("wordStr", keyWord)
                .findAll()
        rows.deleteAllFromRealm()
        realm.commitTransaction()
        realm.close()
        return true
    }

    override fun queryEntries(keyWord: String?, bookName: String): List<Entry> {
        throw RuntimeException("Not implemented yet")
    }

    private fun map(managedEntries: RealmResults<RealmEntry>): List<Entry> {
        val entries = ArrayList<Entry>()
        var i = 0
        while (entries.size < Store.MAX_LENGTH && i < managedEntries.size) {
            val managedEntry = managedEntries[i]
            val entry = Entry()
            // FIXME: should remove !!
            entry.source = managedEntry!!.sourceBook
            entry.wordStr = managedEntry.wordStr
            entry.data = managedEntry.data
            entries.add(entry)
            i++
        }
        return entries
    }
}
